src/package-linker.js

```diff
@@ -423,25 +381,78 @@ export default class PackageLinker {
       }
     }
 
-    for (const loc of possibleExtraneous) {
-      let packageName = path.basename(loc);
-      const scopeName = path.basename(path.dirname(loc));
+    const possibleExtraneous: Set<string> = new Set();
+    const scopedPaths: Set<string> = new Set();
+
+    const findExtraneousFiles = async basePath => {
+      for (const folder of this.config.registryFolders) {
+        const loc = path.resolve(basePath, folder);
+
+        if (await fs.exists(loc)) {
+          const files = await fs.readdir(loc);
+
+          for (const file of files) {
+            const filepath = path.join(loc, file);
+
+            // it's a scope, not a package
+            if (file[0] === '@') {
+              scopedPaths.add(filepath);
+
+              for (const subfile of await fs.readdir(filepath)) {
+                await addPossibleExtraneous(path.join(filepath, subfile));
+              }
+            } else if (file[0] === '.' && file !== '.bin') {
+              if (!(await fs.lstat(filepath)).isDirectory()) {
+                await addPossibleExtraneous(filepath);
+              }
+            } else {
+              await addPossibleExtraneous(filepath);
+            }
+          }
+        }
+      }
+
+      async function addPossibleExtraneous(fpath: string): Promise<void> {
+        let packageName = path.basename(fpath);
+        const scopeName = path.basename(path.dirname(fpath));
+        if (scopeName[0] === `@`) {
+          packageName = `${scopeName}/${packageName}`;
+        }
+        if (
+          (await fs.lstat(fpath)).isSymbolicLink() &&
+          linkTargets.has(packageName) &&
+          linkTargets.get(packageName) === (await fs.realpath(fpath))
+        ) {
+          /* Ignore linked modules. This is now done inside `findExtraneousFiles` itself
+           * rather than afterwards in order to skip recursing into these modules.
+           */
+          copyQueue.delete(fpath);
+          return;
+        }
 
-      if (scopeName[0] === `@`) {
-        packageName = `${scopeName}/${packageName}`;
+        possibleExtraneous.add(fpath);
+        /* Don't recurse into workspace symlinks for two reasons:
```

Comment:
nit: could benefit from a newline between the comment block and the line above it.